src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -365,17 +379,7 @@ export async function run(
     }
     batchesInFlight += 1;
 
-    const checkBatch = submitBatch(
-      t,
-      shaderBuilder,
-      parameterTypes,
-      resultType,
-      batchCases,
-      cfg.inputSource,
-      pipelineCache
-    );
-    checkBatch();
-    t.queue.onSubmittedWorkDone().finally(batchFinishedCallback);
+    processBatch(batchCases);
```

Comment:
You're right! I saw that we had some code to wait on batches of tests and erroneously thought it would wait to end the test until the batches were completed. But it only waits on pending batches if there's more batches to execute.

By adding a `Promise.all()` to the end of the test we do slow down from the numbers I quoted above. Instead on 46s the 4x4 matrix subrtraction test actually takes 80s. So it's _merely ~4x faster_, instead of 6x. My bad. 😉 

Patch incoming.

Reply:
Okay, should be fixed now. Thanks for catching that!